lib-injection/sitecustomize.py

```diff
@@ -1,52 +1,74 @@
 """
-This module when included on the PYTHONPATH will install the ddtrace package
-from the locally available wheels that are included in the image.
+This module when included on the PYTHONPATH will update the PYTHONPATH to point to a directory
+containing the ddtrace package compatible with the current Python version and platform.
 """
 import os
 import sys
 
 
-def _configure_ddtrace():
-    # This import has the same effect as ddtrace-run for the current process.
-    import ddtrace.bootstrap.sitecustomize
+debug_mode = os.environ.get("DD_TRACE_DEBUG") in ("true", "1")
 
-    bootstrap_dir = os.path.abspath(os.path.dirname(ddtrace.bootstrap.sitecustomize.__file__))
-    prev_python_path = os.getenv("PYTHONPATH", "")
-    os.environ["PYTHONPATH"] = "%s%s%s" % (bootstrap_dir, os.path.pathsep, prev_python_path)
 
-    # Also insert the bootstrap dir in the path of the current python process.
-    sys.path.insert(0, bootstrap_dir)
-    print("datadog autoinstrumentation: successfully configured python package", file=sys.stderr)
+def _log(msg, *args, level="info"):
+    """Log a message to stderr.
 
+    This function is provided instead of built-in Python logging since we can't rely on any logger
+    being configured.
+    """
+    if not debug_mode and level == "debug":
+        return
+    print("%s:datadog.autoinstrumentation(pid: %d): " % (level.upper(), os.getpid()) + msg % args, file=sys.stderr)
+
+
+script_dir = os.path.dirname(__file__)
+pkgs_path = os.path.join(script_dir, "ddtrace_pkgs")
+_log("ddtrace_pkgs path is %r" % pkgs_path, level="debug")
+_log("ddtrace_pkgs contents: %r" % os.listdir(pkgs_path), level="debug")
+
+try:
+    import ddtrace
+except ModuleNotFoundError:
+    _log("user-installed ddtrace not found, configuring application to use injection site-packages")
+
+    python_version = ".".join(str(i) for i in sys.version_info[:2])
+    site_pkgs_path = os.path.join(pkgs_path, "site-packages-ddtrace-py%s" % python_version)
```

Comment:
This logic is dependent on how dl_wheels installs and stores packages. Should we move some of this logic to a common utils file or use a constant for  the `site-packages-ddtrace-py-{}`?

Reply:
This is a good callout, there is definitely logic implicitly shared between the modules. Having a shared module would be a good refactor. The tests give good enough validation for now that we can hold off.